File: lesson3/e_2.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
#include <limits>

class Point {
public:
    bool visited;
    double dist;
    std::string from_point;

    Point() : visited(false), dist(std::numeric_limits<double>::infinity()) {}
};

void find_time_to_neighbors(
    const std::string& origin_city,
    const std::string& from_city,
    double prev_time,
    int speed,
    const std::unordered_map<std::string, std::unordered_map<std::string, int>>& roads,
    std::unordered_map<std::pair<std::string, std::string>, double>& times
) {
    if (from_city != "1") {
        for (const auto& road : roads.at(from_city)) {
            const std::string& to_city = road.first;

            if (to_city != from_city && to_city != origin_city) {
                if (times.find({origin_city, to_city}) == times.end()) {
                    times[{origin_city, to_city}] = prev_time + static_cast<double>(road.second) / speed;

                    find_time_to_neighbors(origin_city, to_city, times[{origin_city, to_city}], speed, roads, times);
                }
            }
        }
    }
}

std::unordered_map<std::pair<std::string, std::string>, double> time_searcher(
    const std::unordered_map<std::string, std::unordered_map<std::string, int>>& roads,
    const std::unordered_map<std::string, std::pair<int, int>>& city_info
) {
    std::unordered_map<std::pair<std::string, std::string>, double> times;

    for (const auto& entry : roads) {
        const std::string& from_city = entry.first;

        if (from_city != "1") {
            double prev_time = 0;
            int speed = city_info.at(from_city).second;

            for (const auto& road : entry.second) {
                const std::string& to_city = road.first;

                if (times.find({from_city, to_city}) == times.end()) {
                    times[{from_city, to_city}] = city_info.at(from_city).first + static_cast<double>(road.second) / speed;

                    prev_time = times[{from_city, to_city}];

                    find_time_to_neighbors(from_city, to_city, prev_time, speed, roads, times);
                }
            }
        }
    }

    return times;
}

std::unordered_map<std::string, Point> dijkstraSearchFast(
    const std::unordered_map<std::string, std::unordered_map<std::string, double>>& graph,
    const std::string& point_from
) {
    std::unordered_map<std::string, Point> points_data;
    std::priority_queue<std::pair<double, std::string>, std::vector<std::pair<double, std::string>>, std::greater<>> pts_heap;

    for (const auto& entry : graph) {
        Point new_point;

        if (entry.first == point_from) {
            if (entry.second.find(point_from) == entry.second.end()) {
                new_point.dist = 0;
            } else {
                new_point.dist = entry.second.at(point_from);
            }
        }

        points_data[entry.first] = new_point;
        pts_heap.push({new_point.dist, entry.first});
    }

    bool all_pts_checked = false;
    while (!all_pts_checked) {
        all_pts_checked = true;

        if (!pts_heap.empty()) {
            all_pts_checked = false;

            const std::string& curr_point = pts_heap.top().second;
            pts_heap.pop();

            if (!points_data[curr_point].visited) {
                for (const auto& neighbor : graph.at(curr_point)) {
                    double curr_dist = points_data[curr_point].dist + neighbor.second;

                    if (curr_dist < points_data[neighbor.first].dist) {
                        points_data[neighbor.first].dist = curr_dist;
                        points_data[neighbor.first].from_point = curr_point;
                        pts_heap.push({points_data[neighbor.first].dist, neighbor.first});
                    }
                }

                points_data[curr_point].visited = true;
            }
        } else {
            all_pts_checked = true;
        }
    }

    return points_data;
}

int main() {
    // Read input from file or standard input

    int N;
    std::cin >> N;

    std::unordered_map<std::string, std::unordered_map<std::string, int>> roads;
    std::unordered_map<std::string, std::pair<int, int>> city_info;

    for (int city_id = 1; city_id <= N; ++city_id) {
        int prep_time, speed;
        std::cin >> prep_time >> speed;

        city_info[std::to_string(city_id)] = {prep_time, speed};
    }

    for (int i = 0; i < N - 1; ++i) {
        std::string from_id, to_id;
        int dist;

        std::cin >> from_id >> to_id >> dist;

        roads[from_id][to_id] = dist;
        roads[to_id][from_id] = dist;
    }

    // Perform time search and build the graph
    auto res = time_searcher(roads, city_info);

    std::unordered_map<std::string, std::unordered_map<std::string, double>> graph = {{"1", {}}};
    for (const auto& road : res) {
        const std::string& from_city = road.first.first;
        const std::string& to_city = road.first.second;

        if (graph.find(from_city) == graph.end()) {
            graph[from_city] = {};
        }

        graph[from_city][to_city] = road.second;
    }

    // Find the maximum distance and the corresponding path
    std::string max_pt;
    double max_d = 0;
    std::unordered_map<std::string, Point> max_pts_data;

    for (int city_id = 2; city_id <= N; ++city_id) {
        std::unordered_map<std::string, Point> points_data = dijkstraSearchFast(graph, std::to_string(city_id));

        if (max_d < points_data["1"].dist) {
            max_pts_data = points_data;
            max_d = points_data["1"].dist;
            max_pt = std::to_string(city_id);
        }
    }

    // Restore and print the path
    std::vector<std::string> restored_path = {"1"};
    std::string curr_point = max_pts_data["1"].from_point;

    while (!curr_point.empty()) {
        restored_path.push_back(curr_point);
        curr_point = max_pts_data[curr_point].from_point;
    }

    std::cout << max_d << std::endl;
    for (auto it = restored_path.rbegin(); it != restored_path.rend(); ++it) {
        std::cout << *it << " ";
    }

    return 0;
}

File: lesson3/e_2.py
import math
import heapq


class Point:
    visited = False
    dist = math.inf
    from_point = None


def deikstraSearchFast(
    graph: dict,
    point_from: str
):
    points_data = {}
    max_dist = 0
    point_w_max_dist = None
    
    pts_heap = []
    heapq.heapify(pts_heap)


    for point in graph:
        new_point = Point()
        if point == point_from:
            if not point_from in graph[point]:
                new_point.dist = 0
            else:
                new_point.dist = graph[point][point_from]
        
        points_data[point] = new_point
        heapq.heappush(pts_heap, (new_point.dist, point))


    all_pts_checked = False
    while not all_pts_checked:
        all_pts_checked = True
        if pts_heap:
            all_pts_checked = False
            curr_point = heapq.heappop(pts_heap)[1]
            if not points_data[curr_point].visited:
                

                for neighbor in graph[curr_point]:
                    curr_dist = points_data[curr_point].dist + graph[curr_point][neighbor]

                    if curr_dist < points_data[neighbor].dist:
                        if curr_dist > max_dist:
                            max_dist = curr_dist
                            point_w_max_dist = neighbor
                        
                        points_data[neighbor].dist = curr_dist
                        points_data[neighbor].from_point = curr_point
                        heapq.heappush(pts_heap, (points_data[neighbor].dist, neighbor))

                
                points_data[curr_point].visited = True
            
        else:
            all_pts_checked = True

    return points_data, max_dist, point_w_max_dist



def find_time_to_neighbors(origin_city, from_city, prev_time: int, speed: int, roads: dict, times: dict):
    
    if from_city != '1':
        for to_city in roads[from_city]:
            if to_city != from_city and to_city != origin_city: # check if dead end
                if (origin_city, to_city) not in times:
                    times[(origin_city, to_city)] = prev_time + roads[from_city][to_city] / speed
                    prev_time = times[(origin_city, to_city)]


                    find_time_to_neighbors(origin_city, to_city, prev_time, speed, roads, times)


def time_searcher(roads: dict, city_info: dict, path: dict={}, times: dict={}, end_city:int = '1', visited_end_cities: set=set()):
    # times: key = (from_city, to_city)
    
    neighbors = roads[end_city]

    for neighbor in neighbors:
        if neighbor != end_city and neighbor != '1' and neighbor not in visited_end_cities:
            path[neighbor] = end_city

            # checking times from neighbor to its neighbors on the end_city_side


            # ! use find_time_to_neighbors for both directions

            from_city = neighbor
            to_city = end_city

            prep_time, speed = city_info[from_city]

            if (from_city, to_city) not in times:         
                times[(from_city, to_city)] = prep_time + roads[from_city][to_city] / speed

            prev_time = times[(from_city, to_city)]
            
            if to_city != '1':
                find_time_to_neighbors(neighbor, to_city, prev_time, speed, roads, times)


            # checking times from neighbor to its neighbors on the opposite of end_city_side
            
            neighbors_of_neighbor = roads[neighbor]
            for neighbor_of_neighbor in neighbors_of_neighbor:
                # select those, that were not visited on the way here
                if neighbor_of_neighbor != path[neighbor] and neighbor_of_neighbor != '1':
                    from_city = neighbor
                    to_city = neighbor_of_neighbor

                    if (from_city, to_city) not in times:
                        times[(from_city, to_city)] = prep_time + roads[from_city][to_city] / speed

                    prev_time = times[(from_city, to_city)]

                    find_time_to_neighbors(neighbor, to_city, prev_time, speed, roads, times)

            visited_end_cities.add(neighbor)
            times = time_searcher(roads, city_info, path, times, end_city = neighbor, visited_end_cities=visited_end_cities)

    return times


f = open('/home/oleg/Documents/algo4/lesson3/input.txt')
N = f.readline()
N = int(N)

roads = {}

city_info = {}

for city_id in range( 1,N+1 ):
    prep_time, speed = f.readline().split()
    prep_time = int(prep_time)
    speed = int(speed)

    city_info[str(city_id)] = (prep_time, speed)


for _ in range( N-1 ):
    from_id, to_id, dist = f.readline().split()
    dist = int(dist)


    if from_id not in roads:
        roads[from_id] = {}

    if to_id not in roads[from_id]:
        roads[from_id][to_id] = []

    if to_id not in roads:
        roads[to_id] = {}

    if from_id not in roads[to_id]:
        roads[to_id][from_id] = []


    roads[from_id][to_id] = dist
    roads[to_id][from_id] = dist

res = time_searcher(roads, city_info)

graph = {'1':{}}
for road in res:
    from_city, to_city = road
    
    if from_city not in graph:
        graph[from_city] = {}
    

    graph[from_city][to_city] = res[road]

pt = None
max_d = 0
for city_id in range( 1,N+1 ):
    points_data, max_dist, point_w_max_dist = deikstraSearchFast(graph, str(city_id))
    if max_d < max_dist and point_w_max_dist != '1':
        max_d = max_dist
        pt = point_w_max_dist


print(max_d, pt)
